Add automatic release notes updater

ylemkimon opened the pull request:
- A contributor must include changelog in PR description like:
```markdown
### Brief description of what is fixed or changed
<!-- This will be added to release notes automatically. -->
#### Major changes
* This is a major change
#### Backwards compatibility breaks
* Something has been deprecated!
#### Minor changes
* Tiny, tiny... change

<!-- If changes are trivial or not related to SymPy itself,
leave this section empty and answer 'No' to following question. -->
**Add entry(ies) to the release notes?** No
```
- Travis PR build will fail if changelog is not found or not explicitly skipped
- Travis push build will update release notes of current version

See #10942.
Depends on drdoctr/doctr#276.

ylemkimon committed:
Add automatic release notes updater

Fetches changelogs from PR description and updates release notes on
wiki.

asmeurer commented on .travis.yml:
```diff
@@ -281,6 +281,11 @@ install:
   - set -e
+  - if [[ "${TEST_SPHINX}" == "true" ]]; then
+        cp bin/changelog.py ~/changelog.py;
+        chmod +x ~/changelog.py;
```
Why is the copy necessary? And you should be able to set the permissions to executable in git.

ylemkimon replied:
@asmeurer AFAIK, doctr runs command in the deploy repo, not build repo, and files in the build repo get stashed.

asmeurer replied:
Oh yeah, because it's assumed that any command in the main repo you could just run yourself before running doctr. So I guess this *is* the correct thing to do. We should document this in the doctr docs. I hadn't thought this use-case before (command to run in the deploy repo that lives in the build repo).

asmeurer commented:
This is awesome. We'll probably want to iterate on the exact text and formatting. But first let's get the doctr change merged and released out so we can test this.

asmeurer commented:
Can you add some command line flags to make it easier to test this locally?

ylemkimon committed:
Added testing on local environment

If not on Travis, it will prompt for repository name and PR number,
print parsed changelogs, download release notes from the wiki, and
update the dowloaded file.

ylemkimon commented:
@asmeurer Added testing on local environment. If not on Travis, it will prompt for repository name and PR number, print parsed changelogs, download release notes from the wiki, and update the downloaded file.

ylemkimon committed:
Change deploy repo

ylemkimon committed:
Explicitly include dependency

ylemkimon committed:
Update doctr arguments

ylemkimon commented:
Restarted Travis build per Doctr 1.7.0 release.

ylemkimon commented:
@asmeurer Yes, please add:

```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQCr5R2LqllHrMnP7e3RBijjUV6uP6LdeDolZ72rzQI5A8DYbKuGthV4ZhFSl1aNn1EZ1caWEz6u7zHFfApJTh81JVi2jYMZyhxulclnWrVXYYWp8LJmAOG+aR3XGtMs6m9IMY6eP6ufcKiHu2fkDOZNkAMVjRER1OkwJryZAvgNo6eLUbwYLLpqyeTOBFjVLWf8DMRHvDgCnlHG0NOIty7VXz9ZUVvQYPxgTp2YjPMSQqiO+vPS5CHpfDxnSx6WzTke8X2Kgnxgq5rFJJIkGCFrJp3Up+Oeb2TOisvjZPuypxWMUCRVaz9bD7Wl3+i2EMfLwekEoZrrtHZ0m7TyPu7memK7Hw0VWHJlKcGDK3DNc+jI3Xdg+xaQp8Q7XX4WNJRCezcrJBbhWCnvRsAWMmGLhZmgv5oYyUbCisOH2ioPNScLuw6OMpGRYdtf/hDsIdSbTP5tNwwkIElygqjfNn31TemuIkhSKcAoBw6s2AarP3kidYmys8FkiGI9m+pYwtecY3u6mUJ8UR37MRWGCdEyMtGoH6ZxNDc0SMOMxfSnNZKxpd467r+OPSqOOsPgSfTkRcyOc1kaVkBexlKLw/RHe+h5SJsgg9ecCzaQMVX16tlRk49f2YRnvTnaOPrJigZYIfzWVkjXnNCDe9XLkfe7a23B9Ti8lzkmp3xiRLrKBw== doctr deploy key for sympy/sympy
```

ylemkimon committed:
Add encrypted deploy key and encryption key

ylemkimon committed:
Rename encrypted GitHub deploy key for sympy_doc

asmeurer commented:
I've added it.

ylemkimon committed:
Remove target branch check

Removed target branch check, as it is done by Doctr.

ylemkimon committed:
Fix syntax error

ylemkimon commented:
@asmeurer Thank you.

---
It seems to successfully update release notes: [Travis build log](https://travis-ci.org/sympy/sympy/jobs/306162772), [Release notes(wiki)](https://github.com/sympy/sympy/wiki/Release-Notes-for-1.1.2).

asmeurer commented:
That's great. What happens if no changelog entry is added? Does it fail the tests?

ylemkimon commented:
@asmeurer It will fail if changelog is not found(no template is detected or no entry is added) and not explicitly skipped(with [skip changelog]). See line 75 through 81.

asmeurer commented:
One suggestion: I would define a function like 

```py
def red(text):
    return "\033[31m%s\033[0m" % text
```

and then use that for any major error messages (like `sys.exit(red('Changelog not found! Please add a changelog.'))`). That will make it easier to see in the Travis log.

asmeurer commented:
Since this will have a pretty major effect on the development process, I'd like to get comments from other members of the community before merging it. I'm writing a message to the mailing list.

asmeurer commented on .travis.yml:
```diff
@@ -9,6 +9,8 @@ env:
   - SPLIT="4/4" TEST_SYMPY="true"
   global:
     - secure: "YIEZal9EBTL+fg2YmoZoS8Bvt3eAVUOZjb38CtqpzR2CCSXWoUk35KG23m2NknlY1iKfYJyt7XWBszT/VKOQEbWQq7PIakV4vIByrWacgBxy1x3WC+rZoW7TX+JJiL+y942qIYbMoNMMB8xFpE5RDLSjSecMpFhJJXoafVTvju8="
```
Add a comment that this is for sympy/sympy_doc.

ylemkimon committed:
Add comment for existing sympy_doc deploy encryption key

ylemkimon committed:
Improve console output

Use text color formatting, print the number of changelog entries, and
print run commands.

smichr commented on bin/changelog.py:
```diff
@@ -0,0 +1,175 @@
+    changelogs = defaultdict(list)
+    start = False
+    for line in data.splitlines():
+        if 'This PR changes' in line:
```
I would prefer that it start with `[CHANGELOG START]`. I found that the target text was not always a natural way for me to begin the discussion of the PR. Also, doesn't what you have here treat that opening line like a message to be parsed since it ends with a ":"? Perhaps it would be better to `continue` after detecting the start.

ylemkimon replied:
@smichr `This PR changes` is a newly added level 2 heading on the pull request template.

smichr commented on bin/changelog.py:
```diff
@@ -0,0 +1,175 @@
+            start = True
+        if start:
+            if ':' in line:
+                type, message = line.split(':', 1)
```
There should probably be a warning if the line is too long. The commentary on the change can/should follow a *brief* description of the change and be on a new line:

```
M: brief description of major change
This change was needed for these reasons:
* this
* that
* the other thing
...
```

smichr replied:
Without a warning it is likely that we will get some really long descriptions...like we get when people forget to leave a blank line after a commit title.

ylemkimon replied:
@smichr Actually, it ignores all lines that doesn't start with `(header + ':') for header in HEADERS`, i.e., M:, m:, b:, and n:. I'm not sure there should be warnings on ignored lines, since, as you said, they can be used for comments.

asmeurer commented:
@smichr brings up a good point, which is the verbosity of the changelog entries. I think we should aim toward less verbose entries. If someone wants more information they can click the pull request link. I think having the different headers like `M:` instead of an actual header is good, because it encourages sentences rather than paragraphs.

Another thing, is that this is somewhat duplicate of what already should go in the PR description, but it isn't really clear from the comment. I would make it clearer in the comment text that the stuff in the "this PR changes" actually ends up in the release notes. Maybe the header should just be called "Release notes entry" or something like that to make it abundantly clear.

gxyd commented:
I've a question. Would the final release notes file be a `.rst` file? If it is, then we can reference functions to appropriate documentation via it, ex. if a change is made to `limit` function we can then use formatting :func: \`limits.limit\`.

The format for a changelog entry could be (for PR https://github.com/sympy/sympy/pull/13463):

series:
    Bidirectional limits can now be computed by passing `dir=+-` argument to :func:\`limits.limit\` function :issue:\`11694\` by :user:\`Gaurav Dhingra \<gxyd\>\`

asmeurer commented:
With this everything is markdown, because the pull request descriptions are in markdown.

asmeurer commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,3 +1,8 @@
+
+<!-- Briefly explain what this pull request changes in each line below, with appropriate heading—'M' for major changes, 'm' for minor changes, 'n' for new features, and 'b' for backwards compatibility breaks and deprecations—and a colon(:), e.g., "M: Added changelogs". Write '[skip changelog]' if changes are trivial or not related to SymPy itself. -->
+## This PR changes
+
```
Let's add each heading here, so that people can just fill them out. We might as well use the full words too, like 

````
### Major changes

### Minor changes

### New features

### Backwards compatibility breaks

```

That way people can just fill it out like a template.

gxyd commented:
Ooh, oh. The pull request template was changed, perhaps we will need a merge here.

asmeurer commented:
Also need to update the template and matching per my suggestion above. I'll try to do it later this week if @ylemkimon doesn't get to it. I haven't heard any opposition from the community to the idea so once we clean this up we can merge it and give it a try.

gxyd commented:
Yes, I think that will be good. I'll try to surf through the merged PR's and see if those require updating Release notes. Ping once this is merged.

asmeurer commented:
Yeah we will need to manually add the PRs since 1.1.1 that were already merged.

ylemkimon commented:
Sorry, I've been busy lately. as it was finals season.

@gxyd I think we can enable [Markdown support in Sphinx](http://www.sphinx-doc.org/en/stable/markdown.html) if we import them to the documentation.

@asmeurer In conjunction with #13763, what do you think about:

```markdown
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


### Brief description of what is fixed or changed
<!-- This will be added to changelogs automatically. If changes are trivial
or not related to SymPy itself, please uncomment following line.-->
<!-- [skip changelog] -->
#### Major changes
* 
#### Minor changes
*
#### New features
*
#### Backwards compatibility breaks
*
<!-- [end changelog] -->
### Other comments
```

which renders as:

---
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #13584

### Brief description of what is fixed or changed
<!-- This will be added to changelogs automatically. If changes are trivial
or not related to SymPy itself, please uncomment following line.-->
<!-- [skip changelog] -->
#### Major changes
* This is a major change
#### Minor changes
* Tiny, tiny... change
#### New features
* New feature! Hooray!
#### Backwards compatibility breaks
* Something has been deprecated!
<!-- [end changelog] -->
### Other comments
Hi!

---

gxyd commented:
> Sorry, I've been busy lately. as it was finals season.

Ha ha. :)

asmeurer commented:
> I think we can enable Markdown support in Sphinx if we import them to the documentation.

Woah, is that new? I would definitely like to try that out.

asmeurer commented:
Looking at this, I think "new features" is redundant with "major changes". I guess that comes from the [1.0](https://github.com/sympy/sympy/wiki/Release-Notes-for-1.0) release notes. In those, the "major changes" were *really* major. That was mostly written like that because the release notes were written at the release time, and I wanted to make sure that at least the most important things were in there. For release notes that are written pre-pull request, it makes more sense to just have "major changes" and "minor changes" (and of course, "backwards compatibility breaks"). 

If we want to then summarize at the release what are the key changes, we can do that by hand by adding a new "key changes" section to the top of the notes (so that people can see the most important stuff without having to read the whole thing).

asmeurer commented:
Also we should probably call it "release notes" instead of "changelog", since that's what it's already called on the wiki.

ylemkimon commented:
@asmeurer Updated https://github.com/sympy/sympy/pull/13584#issuecomment-352646581

@gxyd 
>series:
Bidirectional limits can now be computed by passing dir=+- argument to :func:`limits.limit` function :issue:`11694` by :user:`Gaurav Dhingra <gxyd>`

Should we include author's name in the release notes?

gxyd commented:
I think we can drop having author's name, it doesn't add much.

asmeurer commented:
I like that.

asmeurer commented:
I've merged https://github.com/sympy/sympy/pull/13763, so be sure to merge with master.

ylemkimon committed:
Merge branch 'master' into changelog

jksuom commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,9 +1,22 @@
-<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->
+### Brief description of what is fixed or changed
+<!-- This will be added to release notes automatically. If changes are trivial
+or not related to SymPy itself, please uncomment following line.-->
```
Should "uncomment following line" be explained in more detail?

asmeurer replied:
I agree. People might not know how HTML comments work. They might even think the  [] is the comment.

ylemkimon committed:
Update pull request template

- Changed the order of headers
- Instead of using keyword, use plain question to skip release notes.

ylemkimon commented:
@jksuom @asmeurer How about asking a question and parsing the answer, like https://github.com/sympy/sympy/pull/13584#issuecomment-352646581 (https://github.com/sympy/sympy/pull/13584/commits/df227fdafb33c2ae88d2710df619281c27511bfc)?

---

Also, I think pull request template may need following:

1. a notice not to alter pull request template
1. guidelines or instructions on how to write a release notes entry, maybe a link to http://keepachangelog.com/
1. a section for detailed description or explanation

asmeurer commented:
We might have to experiment to see what method of parsing no changelog would be the most natural to new users. 

I like having a changelog guide, although I think we should write our own on the wiki instead of using that site, as they do a few things differently from us and it could cause some confusion.

asmeurer commented:
I'd like to try getting this in soon so we can start using it for the next release after the one I'm working on. I've been working on compiling https://github.com/sympy/sympy/wiki/Release-Notes-for-1.2 from the merged pull requests (over 700) and it's not fun. 

One thing I'm doing for this release that I think is quite helpful is grouping the major changes by submodule (core, solvers, simplify, ...). So I might take a look at this and see how difficult it is to add that. 

But I'm fully expecting this to evolve past the first iteration. The nice thing about the notes going into the wiki is that we can always edit the wiki later to fix things after a PR is merged.

asmeurer commented:
Based on my compiling of https://github.com/sympy/sympy/wiki/Release-Notes-for-1.2, I have some suggestions for this

- There shouldn't be a distinction between major and minor changes. A better organization is per-module. Actual "minor" changes shouldn't be included at all (like test changes, docs changes). We should probably make a wiki page describing the formatting for release notes in more detail and link it.

- The "major" changes should be assembled by hand for each release. These are the big bullet points that are advertised as the biggies for the release. 

- We should include the author name as well (I haven't been doing this for 1.2 because it's too much work for manually compiled notes). 

- Each note should be at most one sentence, with few exceptions. The instructions should make this clear. If each note is a paragraph, then it's no more helpful than just looking through a list of merged PRs. 

- If a PR simply modifies something that is already in the notes, someone should just add a link to that PR to the existing item manually.

asmeurer commented:
I have decided that it will be better to implement this as a separate GitHub bot, which I am working on at https://github.com/sympy/sympy-bot (I have reused some of the code from here).
